spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java

```diff
@@ -0,0 +1,344 @@
+
+		coordinator.subscribe(a, n, actual);
+	}
+
+	@Override
+	public Object scanUnsafe(Attr key) {
+		return null; // no particular key to be represented, still useful in hooks
+	}
+
+	static final class RaceCoordinator<T> implements Subscription, Scannable {
+
+		final FirstNonEmptyEmittingSubscriber<T>[] subscribers;
+
+		volatile boolean cancelled;
+
+		volatile int wip;
+
+		volatile int competingSubscribers;
+
+		@SuppressWarnings("rawtypes")
+		static final AtomicIntegerFieldUpdater<RaceCoordinator> WIP = AtomicIntegerFieldUpdater
+				.newUpdater(RaceCoordinator.class, "wip");
+
+		static final AtomicIntegerFieldUpdater<RaceCoordinator> COMPETING_SUBSCRIBERS = AtomicIntegerFieldUpdater
+				.newUpdater(RaceCoordinator.class, "competingSubscribers");
+
+		@SuppressWarnings("unchecked")
+		RaceCoordinator(int n) {
+			subscribers = new FirstNonEmptyEmittingSubscriber[n];
+			wip = Integer.MIN_VALUE;
+			competingSubscribers = n;
+		}
+
+		@Override
+		public Stream<? extends Scannable> inners() {
+			return Stream.of(subscribers);
+		}
+
+		@Override
+		@Nullable
+		public Object scanUnsafe(Attr key) {
+			if (key == Attr.CANCELLED) {
+				return cancelled;
+			}
+
+			return null;
+		}
+
+		void subscribe(Publisher<? extends T>[] sources, int n,
+				CoreSubscriber<? super T> actual) {
+			FirstNonEmptyEmittingSubscriber<T>[] a = subscribers;
+
+			for (int i = 0; i < n; i++) {
+				a[i] = new FirstNonEmptyEmittingSubscriber<>(actual, this, i);
+			}
+
+			actual.onSubscribe(this);
+
+			for (int i = 0; i < n; i++) {
+				if (cancelled || wip != Integer.MIN_VALUE) {
+					return;
+				}
+
+				Publisher<? extends T> p = sources[i];
+
+				if (p == null) {
+					if (WIP.compareAndSet(this, Integer.MIN_VALUE, -1)) {
+						actual.onError(new NullPointerException(
+								"The " + i + " th Publisher source is null"));
+					}
+					return;
+				}
+
+				p.subscribe(a[i]);
+			}
+
+		}
+
+		@Override
+		public void request(long n) {
+			if (Operators.validate(n)) {
+				int w = wip;
+				if (w >= 0) {
+					subscribers[w].request(n);
+				}
+				else {
+					for (FirstNonEmptyEmittingSubscriber<T> s : subscribers) {
+						s.request(n);
+					}
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			if (cancelled) {
+				return;
+			}
+			cancelled = true;
+
+			int w = wip;
+			if (w >= 0) {
+				subscribers[w].cancel();
+			}
+			else {
+				for (FirstNonEmptyEmittingSubscriber<T> s : subscribers) {
+					s.cancel();
+				}
+			}
+		}
+
+		boolean tryWin(int index) {
+			if (wip == Integer.MIN_VALUE) {
+				if (WIP.compareAndSet(this, Integer.MIN_VALUE, index)) {
+
+					FirstNonEmptyEmittingSubscriber<T>[] a = subscribers;
+					int n = a.length;
+
+					for (int i = 0; i < n; i++) {
+						if (i != index) {
+							a[i].cancel();
+						}
+					}
+
+					return true;
+				}
+			}
+			return false;
+		}
+
+		void resignFromRace() {
+			COMPETING_SUBSCRIBERS.decrementAndGet(this);
```

Comment:
make this method return the result of the `decrementAndGet`, so that caller might directly compare it to `0` and propagate `onComplete`.